Throttle system forms by delaying them slightly

dannyroberts opened the pull request:
## Product Description

https://dimagi-dev.atlassian.net/browse/SAAS-12668

Slow down system-internal submissions (mostly auto case rule updates) when we're above our global target submission rate. This smooths out load on the system, allows a domain's real submissions to share in the bandwidth, and improves stability during the time period (midnight UTC) when these happen in the highest volume.

## Technical Summary
Throttle system forms by delaying them slightly under the same conditions that we would reject a normal submission

## Feature Flag
THROTTLE_SYSTEM_FORMS (Feature Release)

## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below

### Automated test coverage

Added a few tests that just go through the happy path and make sure that it hasn't broken.

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->

### Safety story
I put this on staging to test the case import and other things that use `submit_form_locally`, and test that it works even when it's inducing throttling. Case import and "clean case data" work as normal. "Clean case data" emits one `delay_rather_than_reject` if the flag is enabled, and doesn't emit anything if the flag is not enabled, which is consistent with the flag properly gating the feature. (Case import used to have its own delay which now goes through a custom max_wait in `rate_limit_submission` so that it shows up in metrics, and that is working correctly as well.)

To mitigate risk, the net-new part, throttling all calls to `submit_form_locally` by default, is under a flag. When I roll it out I will also watch the behavior of celery tasks related to auto case close at 7pm CT when they all kick off., as well as metrics related to rate limiting.

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

dannyroberts committed:
Throttle system forms by delaying them slightly

under the same conditions that we would reject a normal submission

millerdev commented on corehq/apps/receiverwrapper/rate_limiter.py:
```diff
@@ -102,19 +106,39 @@ def report_submission_usage(domain):
         duration_tag = bucket_value(timer.duration, [.5, 1, 5, 10, 15], unit='s')
+    elif delay_rather_than_reject:
+        if timer.duration < max_wait:
+            time.sleep(max_wait - timer.duration)
```
Is this a good idea? Seems like it could cause a balloon in gevent workers, which may have negative memory consequences.

dannyroberts replied:
That's a reasonable question. The `submission_rate_limiter.wait` command a few lines above will use time.sleep in sub-second increments in a loop to keep checking up for up to `max_wait` which defaults to 15 seconds, and that's been live for quite some time (since Dec 2019, I believe). It's certainly possible that that causes problems that haven't bubbled up into something we've notice, but I haven't seen it.

This change makes `submit_form_locally` sleep for up to 100 ms (`.1` below) if a "real" submission from that domain would be throttled/rejected. I picked 100 ms because it's on the same order of magnitude as the minimum amount of time we could expect a form submission to take, and so would less-than-double the total amount of time needed to get through N such submissions even if it was being throttled the whole time. The biggest users of `submit_form_locally`, I believe are in case imports and auto case updates, which are both in celery processes that don't use `gevent`, though certainly other places might.

Anyway those are my thoughts on that. Curious to hear what you think.

millerdev replied:
Sounds like you've thought it through. My only remaining concern is we are unlikely to know if this poses a real problem until it gets really bad. Will we remember to look here if this causes the problem I described?

Is `time.sleep` patched to `gevent.sleep` in the context of Gunicorn? If not, should we use `gevent.sleep` instead?

dannyroberts replied:
My understanding is that `gevent.monkey.patch_all()` patches `time.sleep` by default, and we use that default. I double checked the code where we do that and the definition of the `patch_all` function, and it seems to bear that out as well.

I think the potential issue you're describing is mainly a concern on webworkers, if there was some high-volume route that called this function synchronously (as opposed to through celery). We have a metric on our HQ Vitals datadog dashboard that tracks the percent of worker processes in use (split out by mobile_webworkers vs hq_webworkers) so we'd definitely see if it was reaching it's max. We calibrate the machines/process counts so that we basically never hit (or even get that close to) the limit, so it would certainly stand out if it did.

I also in one of my later commits put the "new" part of this PR, rate limiting via a delay within `submit_form_locally` by default, under a flag. That way if there is an issue with that, it's something we can turn off quickly. (There was a pre-existing similar delay in case imports, which I rewrote to go through `submit_form_locally` instead. That's a change in implementation rather than a change in functionality, and I didn't put that change under the flag.)

millerdev commented:
Accidentally approve this draft PR, and now I can't figure out how to withdraw my approval. Please ignore.

dannyroberts committed:
Add very basic test for rate limiting within submit_form_locally

Designed to fail on the current code due to a very obvious typo/bug

======================================================================
ERROR: corehq.apps.receiverwrapper.tests.test_submissions:SubmitFormLocallyRateLimitTest.test_rate_limiting
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/danielroberts/.virtualenvs/cchq/lib/python3.7/site-packages/mock/mock.py", line 1305, in patched
    return func(*args, **keywargs)
  File "/Users/danielroberts/dimagi/commcare-hq/corehq/apps/receiverwrapper/tests/test_submissions.py", line 324, in test_rate_limiting
    submit_form_locally(form_xml, domain=self.domain)
  File "/Users/danielroberts/dimagi/commcare-hq/corehq/apps/receiverwrapper/util.py", line 32, in submit_form_locally
    rate_limit_submission(domain, delay_rather_than_reject=True, rate_limit_submission=.1)
  File "/Users/danielroberts/dimagi/commcare-hq/corehq/util/decorators.py", line 61, in inner
    return fn(*args, **kwargs)
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.7/lib/python3.7/contextlib.py", line 74, in inner
    return func(*args, **kwds)
TypeError: rate_limit_submission() got an unexpected keyword argument 'rate_limit_submission'

dannyroberts committed:
Fix typo-bug I just wrote a test for

dannyroberts committed:
Make case importer take advantage of submit_form_locally for custom throttling

dannyroberts committed:
Add and lightly test THROTTLE_SYSTEM_FORMS FeatureRelease toggle

millerdev commented on corehq/apps/receiverwrapper/util.py:
```diff
@@ -27,7 +29,13 @@ def get_submit_url(domain, app_id=None):
 
 
-def submit_form_locally(instance, domain, **kwargs):
+def submit_form_locally(instance, domain, max_wait=Ellipsis, **kwargs):
```
```suggestion
def submit_form_locally(instance, domain, max_wait=..., **kwargs):
```

millerdev commented on corehq/apps/receiverwrapper/util.py:
```diff
@@ -27,7 +29,13 @@ def get_submit_url(domain, app_id=None):
+def submit_form_locally(instance, domain, max_wait=Ellipsis, **kwargs):
+    # Using Ellipsis here to mean "The caller did not pass in a value",
+    # because the value None has a special meaning
+    if max_wait is Ellipsis:
```
```suggestion
    if max_wait is ...:
```

millerdev commented on corehq/apps/hqcase/utils.py:
```diff
@@ -35,7 +35,7 @@
 def submit_case_blocks(case_blocks, domain, username="system", user_id=None,
                        xmlns=None, attachments=None, form_id=None,
-                       form_extras=None, case_db=None, device_id=None):
+                       form_extras=None, case_db=None, device_id=None, max_wait=Ellipsis):
```
Is `Ellipsis` an alternate constant because `None` will not work here? Probably would be more standard to use the ellipsis literal instead
```suggestion
                       form_extras=None, case_db=None, device_id=None, max_wait=...):
```

millerdev commented on corehq/apps/hqcase/utils.py:
```diff
@@ -50,6 +50,8 @@ def submit_case_blocks(case_blocks, domain, username="system", user_id=None,
     argument. A human recognizable value is recommended outside of test
     code. Example: "auto-close-rule-<GUID>"
+    :param max_wait: Maximum time to allow the process to be delayed if
+    the project is over its submission rate limit.
```
It would be useful if this docstring explained the meaning of the default value since `...` or `Ellipsis` does not make that obvious. It could be as simple as pointing to the dos for `submit_form_locally`.

millerdev reviewed (approved):
No blocking comments.

dannyroberts committed:
Use ... rather than Ellipsis

and improve related docstrings
